dpnp/linalg/dpnp_utils_linalg.py

```diff
@@ -907,6 +951,318 @@ def dpnp_inv(a):
+        elif mode == "complete":
+            return dpnp.identity(
+                m, dtype=res_type, sycl_queue=a_sycl_queue, usm_type=a_usm_type
+            ), dpnp.empty(
+                (m, n),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+        elif mode == "r":
+            return dpnp.empty(
+                (0, n),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+        else:  # mode == "raw"
+            return dpnp.empty(
+                (n, m),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            ), dpnp.empty(
+                (0,),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+
+    a = a.T
+    a_usm_arr = dpnp.get_usm_ndarray(a)
+    a_t = dpnp.empty_like(a, order="C", dtype=res_type)
+
+    # use DPCTL tensor function to fill the matrix array
+    # with content from the input array `a`
+    a_ht_copy_ev, a_copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
+        src=a_usm_arr, dst=a_t.get_array(), sycl_queue=a_sycl_queue
+    )
+
+    tau_h = dpnp.empty(
+        k, dtype=res_type, sycl_queue=a_sycl_queue, usm_type=a_usm_type
+    )
+
+    # Call the LAPACK extension function _geqrf to compute the QR factorization
+    # of a general m x n matrix.
+    ht_geqrf_ev, geqrf_ev = li._geqrf(
+        a_sycl_queue, a_t.get_array(), tau_h.get_array(), [a_copy_ev]
+    )
+
+    ht_geqrf_ev.wait()
+    a_ht_copy_ev.wait()
+
+    if mode == "r":
+        r = a_t[:, :k].transpose()
+        return dpnp.triu(r).astype(res_type, copy=False)
+
+    if mode == "raw":
+        return (
+            a_t.astype(res_type, copy=False),
+            tau_h.astype(res_type, copy=False),
+        )
+
+    # mc is the total number of columns in the q matrix.
+    # In `complete` mode, mc equals the number of rows.
+    # In `reduced` mode, mc is the lesser of the row count or column count.
+    if mode == "complete" and m > n:
+        mc = m
+        q = dpnp.empty(
+            (m, m),
+            dtype=res_type,
+            order="C",
+            sycl_queue=a_sycl_queue,
+            usm_type=a_usm_type,
+        )
+    else:
+        mc = k
+        q = dpnp.empty(
+            (n, m),
+            dtype=res_type,
+            order="C",
+            sycl_queue=a_sycl_queue,
+            usm_type=a_usm_type,
+        )
+    q[:n] = a_t
+
+    # Call the LAPACK extension function _orgqr to generate the real orthogonal matrix
+    # `Q` of the QR factorization
+    ht_orgqr_ev, _ = li._orgqr(
+        a_sycl_queue, m, mc, k, q.get_array(), tau_h.get_array(), [geqrf_ev]
+    )
+
+    ht_orgqr_ev.wait()
+
+    q = q[:mc].transpose()
+    r = a_t[:, :mc].transpose()
+    return (
+        q.astype(dtype=res_type, copy=False),
```

Comment:
the same question regarding `astype` needs and possible improvement though `_triu_inplace`